# Day Four

## Day Overview

### Lecture 4A: Introduction to microbial genomics

**Instructors**: Alexander Herbig

**Abstract**: TBC

### Practical 4B: Genome mapping

**Instructors**: Alexander Herbig and Alina Hiß

**Abstract**: An important step in the reconstruction of full genomic sequences is mapping. Even relatively short genomes usually cannot be sequenced as a single consecutive piece. Instead, millions of short sequence reads are generated from genomic fragments. These reads can be several hundred nucleotides in length but are considerably shorter for ancient DNA (aDNA).

For many applications involving comparative genomics these 'reads' have to be aligned to one or multiple already-reconstructed reference genomes in order to identify differences between the sequenced genome and any given contextual dataset. Aligning millions of short reads to much longer genome sequences in a time-efficient and accurate manner is a bioinformatics challenge for which numerous algorithms and tools have been developed. Each of these programs comes with a variety of parameters that can significantly alter the results and default settings are often not optimal when working with aDNA. Furthermore, read mapping procedures are often part of complex computational genomics pipelines and are therefore not directly applied by many users.

In this session we will take a look at specific challenges during read mapping when dealing with aDNA. We will get an overview of common input and output formats and manually apply a read mapper to aDNA data studying the direct effects of variation in mapping parameters. We will conclude the session with an outlook on genotyping, which is an important follow-up analysis step, that in turn is very relevant for down-stream analyses such as phylogenetics.

### Practical 4C: Genome assembly

**Instructors**: Alexander Hübner and Nikolay Oskolkov

**Abstract**: _De novo_ assembly of ancient metagenomic samples enables the recovery of the genetic information of organisms without requiring any prior knowledge about their genomes. Therefore, this approach is very well suited to study the biological diversity of species that have not been studied well or are simply not known yet.

In this session, we will show you how to prepare your sequencing data and subsequently _de novo_ assemble them. Furthermore, we will then learn how we can actually evaluate what organisms we might have assembled and whether we obtained enough data to reconstruct a whole metagenome-assembled genome. We will particularly focus on the quality assessment of these reconstructed genomes and how we can ensure that we obtained high-quality genomes.

### Roundtable: Databases

**Chairs**: TBC

Arguably, _which_ databases to use are even more important than classifiers when it comes to metagenomics. What is in and not in your database can greatly influence your output results. In this roundtable we will discuss how to select and create databases, and also strategies to get around issues such as computational limitations when constructing and using databases.

## Lecture 4A: Introduction to microbial genomics

### Resources

### Readings

### Questions to think about

### Material

```bash
test
```

## Software

For today's practical sessions, please activate the `microbial-genomics` conda [environment](2022/resources#software-and-data)

```bash
conda activate microbial-genomics
```

## Practical 4B: Genome mapping

### Resources

### Readings

### Questions to think about

### Material

```bash
test
```

## Practical 4C: Genome assembly

### Resources

### Readings

### Questions to think about

### Material

```bash
test
```
