# Instructors

Meet your instructors! This page introduces you to the people guiding you through
ancient metagenomics in the 2022 edition of the summer school.

## Organisers

<div class="avatar">

<div class ="member">
<div class="square"><a href="https://www.jafy.eu/" target="_blank"><img src="assets/images/headshots/FELLOWS_YATES_James.jpg" alt="Avatar" /></a></div>
<p>James Fellows Yates</p>
</div>

<div class ="member">
<div class="square"><a href="http://christinawarinner.com/" target="_blank"><img src="assets/images/headshots/WARINNER_Christina.webp" alt="Avatar" /></a></div>
<p>Prof. Christina Warinner</p>
</div>

</div>

ðŸ‡¬ðŸ‡§ **James Fellows Yates** is an archaeology-trained biomolecular archaeologist and convert to palaeogenomics, and is recently pivoting to bioinformatics. He specialises in ancient metagenomics analysis, generating tools and high-throughput approaches and high-quality pipelines for validating and analysing ancient (oral) microbiomes and palaeogenomic data.

ðŸ‡ºðŸ‡¸ **Christina Warinner** is Group Leader of Microbiome Sciences at the Max Planck Institute for Evolutionary Anthropology in Leipzig, Germany, and Associate Professor of Anthropology at Harvard University. She serves on the Leadership Team of the Max Planck-Harvard Research Center for the Archaeoscience of the Ancient Mediterranean (MHAAM), and is a Professor in the Faculty of Biological Sciences at Friedrich Schiller University in Jena, Germany. Her research focuses on the use of metagenomics and paleoproteomics to better understand past human diet, health, and the evolution of the human microbiome.

## Instructors

<div class="avatar">

<div class ="member">
<div class="square"><a href="https://twitter.com/aidaanva" target="_blank"><img src="assets/images/headshots/ANDRADES_VALTUENA_Aida.jpg" alt="Avatar" /></a></div>
<p>Aida Andrades ValtueÃ±a</p>
</div>

<div class ="member">
<div class="square"><a href="https://www.eva.mpg.de/archaeogenetics/research-groups/computational-pathogenomics/" target="_blank"><img src="assets/images/headshots/HERBIG_Alexander.webp" alt="Avatar" /></a></div>
<p>Alexander Herbig</p>
</div>

<div class ="member">
<div class="square"><a href="https://twitter.com/alexhbnr" target="_blank"><img src="assets/images/headshots/HUEBNER_Alex.jpg" alt="Avatar" /></a></div>
<p>Alex HÃ¼bner</p>
</div>

<div class ="member">
<div class="square"><a href="https://www.eva.mpg.de/archaeogenetics/research-groups/computational-pathogenomics/" target="_blank"><img src="assets/images/headshots/HISS_Alina.jpg" alt="Avatar" /></a></div>
<p>Alina HiÃŸ</p>
</div>

<div class ="member">
<div class="square"><a href="https://www.researchgate.net/profile/Arthur-Kocher" target="_blank"><img src="assets/images/headshots/KOCHER_Arthur.jpg" alt="Avatar" /></a></div>
<p>Arthur Kocher</p>
</div>

<div class ="member">
<div class="square"><a href="https://nevrome.de" target="_blank"><img src="assets/images/headshots/SCHMID_Clemens.JPG" alt="Avatar" /></a></div>
<p>Clemens Schmid</p>
</div>

<div class ="member">
<div class="square"><a href="https://twitter.com/FzzyToothSweatr" target="_blank"><img src="assets/images/headshots/VELSKO_Irina.jpeg" alt="Avatar" /></a></div>
<p>Irina Velsko</p>
</div>

<div class ="member">
<div class="square"><a href="https://maximeborry.com/" target="_blank"><img src="assets/images/headshots/BORRY_Maxime.png" alt="Avatar" /></a></div>
<p>Maxime Borry</p>
</div>

<div class ="member">
<div class="square"><a href="https://scholar.harvard.edu/megan-michel/home" target="_blank"><img src="assets/images/headshots/MICHEL_Megan.jpg" alt="Avatar" /></a></div>
<p>Megan Michel</p>
</div>

<div class ="member">
<div class="square"><a href="https://twitter.com/TCLamnidis" target="_blank"><img src="assets/images/headshots/LAMNIDIS_Thiseas.jpg" alt="Avatar" /></a></div>
<p>Thiseas C. Lamnidis</p>
</div>

</div>

ðŸ‡ªðŸ‡¸ **Aida Andrades ValtueÃ±a** is a geneticist interested in pathogen evolution, with particular interest in prehistoric pathogens. She has been exploring new methods to analyse ancient pathogen data to understand their past function and ecology to inform models of pathogen emergence.

ðŸ‡©ðŸ‡ª **Alexander Herbig** is a bioinformatician and group leader for Computational Pathogenomics at the Max Planck Institute for Evolutionary Anthropology. His main interest is in studying the evolution of human pathogens and in methods development for pathogen detection and bacterial genomics.

ðŸ‡©ðŸ‡ª **Alex HÃ¼bner** is a computational biologist, who originally studied biotechnology, before switching to evolutionary biology during his PhD. For his postdoc in the Warinner lab, he focuses on investigating whether new methods in the field of modern metagenomics can be directly applied to ancient DNA data. Here, he is particularly interested in the _de novo_ assembly of ancient metagenomic sequencing data and the subsequent analysis of its results.

ðŸ‡©ðŸ‡ª **Alina Hiss** is a PhD student in the Computational Pathogenomics group at the Max Planck Institute for Evolutionary Anthropology. She is interested in the evolution of human pathogens and working on material from the Carpathian basin to gain insights about the presence and spread of pathogens in the region during the Early Medieval period.

ðŸ‡«ðŸ‡· **Arthur Kocher** initially trained as a veterinarian. He then pursued a PhD in the field of disease ecology, during which he studied the impact of biodiversity changes on the transmission of zoonotic diseases using molecular tools such as DNA metabarcoding. During his Post-Docs, he extended his research focus to evolutionary aspects of pathogens, which he currently investigates using ancient genomic data and Bayesian phylogenetics.

ðŸ‡©ðŸ‡ª **Clemens Schmid** is a computational archaeologist pursuing a PhD in the group of Stephan Schiffels at the department of Archaeogenetics at the Max Planck Institute for Evolutionary Anthropology. He is trained both in archaeology and computer science and currently develops computational methods for the spatiotemporal co-analysis of archaeological and ancient genomic data. He worked in research projects on the European Neolithic, Copper and Bronze age and maintains research software in R, C++ and Haskell.

ðŸ‡ºðŸ‡¸ **Irina Velsko** is a postdoc in the Microbiome group of the department of Archaeogenetics at the Max Planck Institute for Evolutionary Anthropology. She did her PhD work on oral microbiology and immunology of the living, and now works on oral microbiomes of the living and the dead. Her work focuses on the evolution and ecology of dental plaque biofilms, both modern and ancient, and the complex interplay between microbiomes and their hosts.

ðŸ‡«ðŸ‡· **Maxime Borry** is a doctoral researcher in bioinformatics at the Max Planck Institute for Evolutionary Anthropology in Germany. After an undergraduate degree in biology, he first decided to dive in the studies of population and our environment. After realizing he wanted to focus more on the statistical part and the data analysis, he switched back to a Bioinformatics MSc. In his PhD, he focuses on developing new validation tools and applying machine learning techniques to ancient metagenomic samples.

ðŸ‡ºðŸ‡¸ **Megan Michel** is a PhD student jointly affiliated with the Archaeogenetics Department at the Max Planck Institute for Evolutionary Anthropology and the Human Evolutionary Biology Department at Harvard University. Her research focuses on using computational genomic analyses to understand how pathogens have co-evolved with their hosts over the course of human history.

ðŸ‡¬ðŸ‡· **Thiseas Lamnidis** is a human population geneticist interested in European population history after the Bronze Age. To gain the required resolution to differentiate between Iron Age European populations, he is developing analytical methods based on the sharing of rare variation between individuals. He has also contributed to pipelines that streamline the processing and analysis of genetic data in a reproducible manner, while also facilitating dissemination of information among interdisciplinary colleagues.

<style>
.member {
  width: 7rem;
  text-align: center;
}

.square {
  display: inline-block;
  width: 6rem;
  height: 6rem;
  margin: auto;
  background-color: #fff;
}

.square img {
  opacity: 1;
  -webkit-transition: 0.3s ease-in-out;
  transition: 0.3s ease-in-out;
}

.square:hover img {
  opacity: 0.5;
}

.avatar {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
}

.avatar img {
  border-radius: 50%;
  width: 6rem;
  height: 6rem;
  object-fit: cover;
  display: block;
  margin: auto;
}

.member p {
  text-align: center;
  font-size: 0.7rem;
  margin-bottom: 0;
  display: block;
}

.member p:first-of-type {
  font-size: 0.8rem;
  color: #000;
  font-weight: bold;
}
</style>
