# Introduction to Ancient Metagenomics Summer School - 2022

## What

The **Ancient Metagenomics Summer School** will be a 5 day **practical** workshop (block praktikum). Attendees who participate in the entire course will be eligible for 1.5 ECTS points (40 hours of instruction and **practicals**, 5 hours of coursework).

The course is taught by scientific staff at the Max Planck Institute for Evolutionary Anthropology and supported by the Werner Siemens Foundation, in collaboration with the Max Planck Harvard Center for the Archaeoscience of the Ancient Mediterranean (MHAAM), the Leibniz Institute for Natural Product Research and Infection Biology, and the Jena School for Microbial Communications. The course coordinators are Prof. Dr. Christina Warinner (christina_warinner@eva_mpg.de) and James Fellows Yates (james_fellows_yates@eva.mpg.de).

## Who

The course is aimed at masters students and early-stage PhD students, to a maximum of 25 participants.

You can meet the instructors [here](/2022/instructors.md).

## Where

This summer school will be held virtually on the online platform gather.town. More details to follow.

## When

The summer school will be provisionally held between 1st-5th August, 09:00-17:00 CEST (Berlin Time).

Note the following additional dates:

- **Applications open**: 29th April 2022
- **Applications close**: 1st June 2022
- **Decision notifications**: 15th June 2022

## What

You will need access to the internet, and a terminal of some form (see this editions page [Resources](2022/resources) for suggestions)

## Contents Overview

### DAY 1

- **Lecture** Introduction to NGS data: _James Fellows Yates_
  - What is NGS sequencing and what does its data look like?
- **Practical** BareBonesBash 1: _Aida Andrades Valtue単a and Thiseas Lamnidis_
  - Introduction to the UNIX command line.
- **Practical** BareBonesBash 2: _Aida Andrades Valtue単a and Thiseas Lamnidis_
  - How to navigate and batch-manipulate files.
- **Round table** Introductions
  - Welcome - let's get to know eachother

### DAY 2

- **Lecture** Introduction to ancient DNA: _Christina Warinner_
  - How to identify and handle DNA from ancient contexts
- **Practical** Bytesize git: _James Fellows Yates and Megan Michel_
  - What is git and github? How to clone a repository? What does commit/pull/push mean?
- **Practical** ancientMetagenomeDir: _James Fellows Yates and Megan Michel_
  - Where to get ancient metagenomic comparative data, and why is metadata important?
  - Git practice
- **Practical** nf-core/eager: _James Fellows Yates and Megan Michel_
  - How do I quality control NGS aDNA data? How do I remove adapters and contamination? How to perform taxonomic profiling?
- **Round table** Project organisation
  - How to structure my project analysis folders? How to reproducibly document my work?

### DAY 3

- **Lecture** Introduction to Metagenomics: _TBA_
  - What are the questions tackled in metagenomics? What are the challenges?
- **Practical** Introduction to R and the tidyverse: _Clemens Schmid and Nikolay Oskolkov_
  - How to read/write files and run statistical analyses? How to make figures with ggplot2
- **Practical** Taxonomic Profiling, OTU Tables and Visualisation: _Maxime Borry and Irina Velsko_
  - R practice. How to normalise OTU tables? How do I remove contamination?
- **Round table** Taxonomic Classifiers
  - How do I pick a taxonomic classifier? What is best for aDNA?

### DAY 4

- **Lecture** Introduction to microbial genomics: _Alexander Herbig_
  - How can we learn from studying the genomes of microbes? What questions can we ask?
- **Practical** Genome mapping: _Alexander Herbig and Alina Hiss_
  - Why and how do we map against a reference? What parameters are important for aDNA?
- **Practical** Genome assembly: _Alexander H端bner and Nikolay Oskolkov_
  - De novo methods for obtaining metagenomically assembled genomes (MAGs). What are the best pipelines and parameters for ancient DNA?
- **Roundtable** Databases
  - Why are databases so important? What impact does database selection have on my results? What pitfalls should I watch out for?

### DAY 5

- **Lecture** Evolutionary Biology: TBA
  - How have microbes and microbial communities evolved and changed through time? How does this inform our current understanding of the relationships among microbes?
- **Practical** Phylogenomics: _Aida Andrades Valtue単a and Arthur Kocher_
  - How to perform phylogenetic analysis. What to consider when dealing with low coverage data
- **Practical** Functional Genomics: _Irina Velsko_
  - How can we infer genomic function from sequences? What sort of analysis can we do?
- **Round table** Workshop recap
