# Resources

This edition of the summerschool will feature heavy use of the command-line.
We ask that you have access to a modern web browser, and/or a terminal system.

## Accessing Compute Nodes

## GitHub account

Although the skills learnt in this workshop will be transferable between different platforms (e.g. GitLab, BitBucket etc.). This workshop will heavily use GitHub for training in `git`. Participants will need to create an account at [github.com](https://github.com) if they have not already done so.

## Software and Data

We will be using the following software during the summer school.
While you will not need to install these yourself for the summer school, we provide links here to the software used for installation and usage in your own time.

### Software

- [git](https://git-scm.com/) (normally installed by default on all UNIX based operating systems e.g. Linux, OSX)
- [R](https://www.r-project.org/)
- [tidyverse](https://www.tidyverse.org/)
- [nextflow](https://www.nextflow.io/)
- [conda](https://docs.conda.io/en/latest/miniconda.html)
- [nf-core/eager](https://nf-co.re/eager)

### Data

- [AncientMetagenomeDir](https://github.com/SPAAM-community/AncientMetagenomeDir/)
- [ENA](https://www.ebi.ac.uk/ena/browser/view/)

## Literature

We also suggest reading the following literature

Warinner, Christina, Alexander Herbig, Allison Mann, James A. Fellows Yates, Clemens L. Weiß, Hernán A. Burbano, Ludovic Orlando, and Johannes Krause. 2017. “A Robust Framework for Microbial Archaeology.” Annual Review of Genomics and Human Genetics 18 (August): 321–56. [https://doi.org/10.1146/annurev-genom-091416-035526](https://doi.org/10.1146/annurev-genom-091416-035526).

Der Sarkissian, Clio, Irina M. Velsko, Anna K. Fotakis, Åshild J. Vågene, Alexander Hübner, and James A. Fellows Yates. 2021. “Ancient Metagenomic Studies: Considerations for the Wider Scientific Community.” mSystems 6 (6): e0131521. [https://doi.org/10.1128/msystems.01315-21](https://doi.org/10.1128/msystems.01315-21).

Velsko, Irina M., Laurent A. F. Frantz, Alexander Herbig, Greger Larson, and Christina Warinner. 2018. “Selection of Appropriate Metagenome Taxonomic Classifiers for Ancient Microbiome Research.” mSystems 3 (4). https://doi.org/10.1128/mSystems.00080-18.

Arizmendi Cárdenas, Yami Ommar, Samuel Neuenschwander, and Anna-Sapfo Malaspinas. 2022. “Benchmarking Metagenomics Classifiers on Ancient Viral DNA: A Simulation Study.” PeerJ 10 (March): e12784. https://doi.org/10.7717/peerj.12784.
