# Schedule

## Overview

The PROVISIONAL schedule for the summer school is as follows:

| **Time**    | **Session Type** | **Day 1**                                                               | **Day 2**                                                                       | **Day 3**                                         | **Day 4**                          | **Day 5**                                                                           |
| ----------- | ---------------- | ----------------------------------------------------------------------- | ------------------------------------------------------------------------------- | ------------------------------------------------- | ---------------------------------- | ----------------------------------------------------------------------------------- |
| 09:00-10:15 | Theoretical      | Welcome & Introduction to NGS data                                      | Introduction to ancient DNA                                                     | Introduction to metagenomics                      | Introduction to microbial genomics | - Decontamination (30m)<br>- General Authentication Lecture and Practical (1h) <br> |
| 10:15-10:30 |                  | Break                                                                   | Break                                                                           | Break                                             | Break                              | Break                                                                               |
| 10:30-12:30 | Practical        | BareBonesBash                                                           | Introduction to R and the tidyverse & Introduction to Python and Pandas         | Taxonomic Profiling, OTU Tables and Visualisation | Genome assembly                    | - Introduction to eDNA (1h) <br> - Authentication at Scale Lecture (30m)            |
| 12:30-13:30 |                  | Lunch                                                                   | Lunch                                                                           | Lunch                                             | Lunch                              | Lunch                                                                               |
| 13:30-15:30 | Practical        | Introduction to R and the tidyverse & Introduction to Python and Pandas | - Introduction to Git(Hub) (1h)<br>- Introduction to AncientMetagenomeDir (1hr) | Genome mapping                                    | Phylogenomics                      | - Authentication at Scale Practical (1h) <br>- Ancient Metagenomic Pipelines (1h)   |
| 15:30-15:45 |                  | Break                                                                   | Break                                                                           | Break                                             | Break                              |                                                                                     |
| 15:45-17:00 | Roundtable       | TBC                                                                     | TBC Activity                                                                    | TBC                                               | Instructor Office Hours            | Dinner (Leipzig Only)                                                               |

Lectures and Practical sessions are typically 2 hours, and round-table discussions 1 hour.

\* All times are in CEST (Leipzig, Germany)!

## Contents

### DAY 1

- **Lecture** Introduction to NGS data: _James Fellows Yates_
  - What is NGS sequencing and what does its data look like?
- **Practical** BareBonesBash: _Aida Andrades Valtueña and Thiseas Lamnidis_
  - Introduction to the UNIX command line.
- **Practical** Introduction to R and the tidyverse / Introduction to Python and Pandas: _Clemens Schmid and Keri Burge_ / _Rémi Denise and Freya Steinhagen_
  - How to read/write files and run statistical analyses? How to make figures with ggplot2/plotnine
  - (Note: Parallel sessions participants can choose which to attend based on their prior experience)
- **Round table** TBC

### DAY 2

- **Lecture** Introduction to ancient DNA: _Christina Warinner_
  - How to identify and handle DNA from ancient contexts
- **Practical** Introduction to R and the tidyverse / Introduction to Python and Pandas: _Clemens Schmid and Keri Burge_ / _Rémi Denise and Freya Steinhagen_
  - How to read/write files and run statistical analyses? How to make figures with ggplot2/plotnine
  - (Note: Parallel sessions participants can choose which to attend based on their prior experience)
- **Practical** Bytesize git: _James Fellows Yates and Megan Michal_
  - What is git and github? How to clone a repository? What does commit/pull/push mean?
- **Practical** AncientMetagenomeDir: _James Fellows Yates and Megan Michal_
  - Where to get ancient metagenomic comparative data, and why is metadata important?
  - Git practice
- **Round table** TBC

### DAY 3

- **Lecture** Introduction to Metagenomics: _Christina Warinner_
  - What are the questions tackled in metagenomics? What are the challenges?
- **Practical** Taxonomic Profiling, OTU Tables and Visualisation: _Keri Burge and Rémi Denise_
  - Python practice. How to normalise OTU tables and perform microbial ecology analyses.
- **Practical** Genome mapping: _Tessa Zeibig and Alexander Herbig_
  - Why and how do we map against a reference? What parameters are important for aDNA?
- **Round table** TBC

### DAY 4

- **Lecture** Introduction to microbial genomics: _Meriam Guellil_
  - How can we learn from studying the genomes of microbes? What questions can we ask?
- **Practical** Genome assembly: _Alex Hübner and Jasmin Frangenberg_
  - De novo methods for obtaining metagenomically assembled genomes (MAGs). What are the best pipelines and parameters for ancient DNA?
- **Practical** Phylogenomics: _Arthur Kocher and Aida Andrades Valtueña_
  - How to perform phylogenetic analysis. What to consider when dealing with low coverage data
- **Roundtable** Instructor Office Hours: Ask the instructors anything about you learnt during the week or to get advice on your own research!

### DAY 5

- **Lecture** Introduction to eDNA: _Vilma Pérez_
  - How have microbes and microbial communities evolved and changed through time? How does this inform our current understanding of the relationships among microbes?
- **Practical** Authentication and Decontamination (microb(iom)es): _Nikolay Oskolkov, Freya Steinhagen, and James Fellows Yates_
  - How can I validate my DNA is ancient? What sort of analyses can convince ourselves and a reviewer we have good preservation?
- **Practical** Authentication and Decontamination (eukaryotes): _Giulia Zampirolo and James Fellows Yates_
  - How can I validate my DNA is ancient? What sort of analyses can convince ourselves and a reviewer we have good preservation?
- **Practical** Ancient Metagenomic Pipelines: _Nikolay Oskolkov, James Fellows Yates_
  - What pipelines exist for specifically ancient DNA and ancient metagenomic data?
- **Round table** (Leipzig only) Dinner for Students and Instructors.
